### PR TITLE
Kernel: The UBSAN story in the aarch64 Kernel

### PR DESCRIPTION
**Pull request by FireFox317**
# UBSAN story
This series of commits began with me checking out if disabling UBSAN on the aarch64 Kernel build actually worked. You know what, it did not work, and exceptions were generated. Next up, these exceptions were not handled, since the proper handlers were not installed.

Without UBSAN the compiler is now able to optimize more, which caused multiple things to break and generate more exceptions. One of which was a data abort, caused by a misaligned memory access. All these exceptions and compiler errors are fixed by these commits.

Next up, I realized that disabling UBSAN, significantly reduced the size of the Kernel binary by 73%(!).

I decided it would be a good idea to disable UBSAN for now for the aarch64 build, since enabling it shadowed certain issues that were in the code, but were never actually discovered because of UBSAN. 
# Miscellaneous
Also adds two fixes for the recent GCC 12.1.0 merge (#13954)
- Disabling vectorization for the aarch64 build
- Fixing capitalization for MiniStdLib in Kernel CMake file.

<details><summary>Spoiler for my next aarch64 PR</summary> I'm gonna work on interrupt support :) </details>

### CODE REVIEW

**Review by Hendiadyoin1 — COMMENTED** [review]
Great Work,
Just some optional nits

File: Kernel/Arch/aarch64/Exceptions.cpp [hunk 1]
```diff
@@ -46,6 +46,10 @@ static void drop_to_el1()
     hypervisor_configuration_register_el2.RW = 1; // EL1 to use 64-bit mode
     Aarch64::HCR_EL2::write(hypervisor_configuration_register_el2);
 
+    // Setup initial exception stack
```
Comment (nico) [inline Kernel/Arch/aarch64/Exceptions.cpp:49]: Nit: "Setup" is a noun, "set up" (with space) is the verb. (Also in commit desc)

File: Kernel/CMakeLists.txt [hunk 2]
```diff
@@ -460,9 +460,13 @@ endif()
 add_compile_options(-fsigned-char)
 add_compile_options(-Wno-unknown-warning-option -Wvla -Wnull-dereference)
 add_compile_options(-fno-rtti -ffreestanding -fbuiltin)
+
 if ("${SERENITY_ARCH}" STREQUAL "i686" OR "${SERENITY_ARCH}" STREQUAL "x86_64")
     add_compile_options(-mno-80387 -mno-mmx -mno-sse -mno-sse2)
+elseif("${SERENITY_ARCH}" STREQUAL "aarch64")
+    add_compile_options(-fno-tree-vectorize)
```
[the remark this answers is not in the document]
Comment (BertalanD) [reply Kernel/CMakeLists.txt:467]: +1

The `-mgeneral-regs-only` flag works with Clang as well.

**Comment by kleinesfilmroellchen** [comment]
First commit message should be present tense

File: Kernel/Arch/aarch64/Registers.h [hunk 3]
```diff
@@ -14,22 +14,25 @@ namespace Kernel::Aarch64 {
...
-    int TGran64 : 4;
-    int TGran4 : 4;
-    int TGran16_2 : 4;
-    int TGran64_2 : 4;
-    int TGran4_2 : 4;
-    int ExS : 4;
-    int RES0 : 8;
-    int FGT : 4;
-    int ECV : 4;
+union ID_AA64MMFR0_EL1 {
```
[the remark this answers is not in the document]
Comment (FireFox317) [reply Kernel/Arch/aarch64/Registers.h:17]: This does indeed also work, and apparently also forces the struct to be passed in a register. Since (for now at least) I don't need access to both the fields and the raw value, a struct+union is not necessary indeed. Will convert to `alignas(u64)`.

File: Kernel/Arch/aarch64/Registers.h [hunk 4]
```diff
@@ -491,4 +491,28 @@ union MAIR_EL1 {
...
+    u64 value;
+    struct {
+        u32 ISS : 25;
+        u32 IL : 1;
+        u32 EC : 6;
+        u32 ISS2 : 5;
+        u32 _reserved0 : 27;
+    };
+
+    static inline ESR_EL1 read()
```
[the remark this answers is not in the document]
Comment (FireFox317) [reply Kernel/Arch/aarch64/Registers.h:506]: I just copied the current style in the code.

File: Kernel/CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (FireFox317) [reply Kernel/CMakeLists.txt:467]: Ah thanks for the suggestion! That is definitely better! Will change to flag.

File: Kernel/Arch/aarch64/Exceptions.cpp (hunk 1 above)
Comment (FireFox317) [reply Kernel/Arch/aarch64/Exceptions.cpp]: Fixed :)

**Comment by FireFox317** [comment]
Thanks for all the reviews! Appreciate it!

I changed the register definitions to use the `alignas` specifier, by which the compiler knows it can put the struct into a 64bit register. This achieves the same goal, but is better. Thanks @Hendiadyoin1 for the suggestion.

Also the `-mgeneral-regs-only` is now used!

File: Kernel/Arch/aarch64/init.cpp [hunk 5]
```diff
@@ -45,7 +46,12 @@ extern "C" void exception_common(TrapFrame const* const trap_frame)
         dbgln("elr_el1: {:x}", trap_frame->elr_el1);
         dbgln("tpidr_el1: {:x}", trap_frame->tpidr_el1);
         dbgln("sp_el0: {:x}", trap_frame->sp_el0);
+
+        auto esr_el1 = Kernel::Aarch64::ESR_EL1::read();
+        dbgln("esr_el1: EC(0b{:b}) IL(0b{:b}) ISS(0b{:b}) ISS2(0b{:b})", esr_el1.EC, esr_el1.IL, esr_el1.ISS, esr_el1.ISS2);
```
Comment (Hendiadyoin1) [inline Kernel/Arch/aarch64/init.cpp:51]: We can let the formatter handle the `0b` prefix
```suggestion
        dbgln("esr_el1: EC({:#b}) IL({:#b}) ISS({:#b}) ISS2({:#b})", esr_el1.EC, esr_el1.IL, esr_el1.ISS, esr_el1.ISS2);
```

Comment (FireFox317) [reply Kernel/Arch/aarch64/init.cpp]: Ah thanks, wasn't aware of that. Fixed!

File: Kernel/CMakeLists.txt [hunk 6]
```diff
@@ -559,6 +559,15 @@ endif()
 
 if ("${SERENITY_ARCH}" STREQUAL "aarch64")
     add_compile_options(-fno-threadsafe-statics)
+
+    # This is needed on aarch64, since the compiler will otherwise assume that
+    # unaligned memory accesses are handled by the kernel/hardware, however
+    # we enforce everything to be properly aligned. Thus we need this option
+    # to make sure that the compiler actually aligns all the data properly.
```
Comment (nico) [inline Kernel/CMakeLists.txt:573]: FWIW, this sounds vaguely fishy to me. If we make sure everything is aligned, then why does the compiler have to do anything?

This comment isn't based in reality in any form really – I'm not sure what this flag does. But I know a few similar flags, and what they often mean is "assume user code is not aligned, but the system requires it, so insert fixups to make certain things are aligned".

What I think we want is to make our user code all aligned, so that even though the system doesn't handle unaligned accesses, we don't have any unaligned accesses since all our code and data is already aligned correctly.

(But again, this comment is based on feeling and experience with similar flags, not with this particular flag. So it's possible I'm off here.)

**Comment by nico** [comment]
Very cool. I have three comments here, none blocking, most of them a bit soapboxy (sorry!):

1. As said above, the `-mstrict-align` flag feels a bit fishy to me
2. Having a way to disable ubsan to find issues hidden by ubsan seems like a useful thing to me. I'm not sure the default of that setting should be tied to aarch64 though – having different defaults for different archs tends to be confusing (for settings that aren't fundamentally tied to the ISA). So I'd weakly prefer to have a cmake option for having ubsan for kernel on/off, having it default to on, and people working on aarch64 (or others) can turn it off if they want. (But this is just my opinion. And during early bringup anything's fair game, so it's not even all that strong of an opinion.)
3. Mostly unrelated to this patch, but explicit bit shuffling operations feel a bit more reliable to me than bitfields.

Thank you so much for pushing this port forward! Really cool work :)

File: Kernel/CMakeLists.txt (hunk 6 above)
Comment (FireFox317) [reply Kernel/CMakeLists.txt:573]: Yes, I agree that we should make sure that the code itself actually make sure everything is properly aligned. This is where the misalignment is happening:

```
   0x00000000000939d8 <+232>:   b.ne    0x939c8 <Kernel::init()+216>  // b.any
   0x00000000000939dc <+236>:   bl      0xb7ae0 <_ZN6Kernel24load_kernel_symbol_tableEv>
   0x00000000000939e0 <+240>:   add     x23, sp, #0x70
   0x00000000000939e4 <+244>:   mov     x0, x23
   0x00000000000939e8 <+248>:   bl      0x92940 <_ZN6Kernel7Mailbox13MessageHeaderC2Ev>
   0x00000000000939ec <+252>:   add     x0, sp, #0x78
   0x00000000000939f0 <+256>:   mov     w2, #0x4                        // #4
   0x00000000000939f4 <+260>:   mov     w1, #0x1                        // #1
   0x00000000000939f8 <+264>:   bl      0x92920 <_ZN6Kernel7Mailbox7MessageC2Ejj>
   0x00000000000939fc <+268>:   stur    xzr, [sp, #132] <--------------------------------------------------------------------------
   0x0000000000093a00 <+272>:   mov     w19, #0xffffffff                // #-1
   0x0000000000093a04 <+276>:   bl      0x92970 <_ZN6Kernel7Mailbox3theEv>
   0x0000000000093a08 <+280>:   add     x24, sp, #0x98
   0x0000000000093a0c <+284>:   mov     x1, x23
   0x0000000000093a10 <+288>:   mov     w2, #0x20                       // #32
   0x0000000000093a14 <+292>:   bl      0x92980 <_ZNK6Kernel7Mailbox10send_queueEPvj>
```

Which is in this function:
https://github.com/SerenityOS/serenity/blob/857a767ab49b6813707b81d93fab81d524cca574/Kernel/Arch/aarch64/init.cpp#L138-L151

It's pretty janky, because the `MessageHeader` and `MessageTail` are just normal classes, no packed structs or anything. So a think something is going wrong here. For now UBSAN is enabled by default, which will get rid of the exception, but this code should still be fixed sometime.

**Comment by FireFox317** [comment]
> 1. As said above, the `-mstrict-align` flag feels a bit fishy to me

TLDR: I think something is wrong with the code indeed. For now KUBSAN is enabled by default, so this exception won't happen. But the code should still be fixed sometime.

> 2. Having a way to disable ubsan to find issues hidden by ubsan seems like a useful thing to me. I'm not sure the default of that setting should be tied to aarch64 though – having different defaults for different archs tends to be confusing (for settings that aren't fundamentally tied to the ISA). So I'd weakly prefer to have a cmake option for having ubsan for kernel on/off, having it default to on, and people working on aarch64 (or others) can turn it off if they want. (But this is just my opinion. And during early bringup anything's fair game, so it's not even all that strong of an opinion.)

Yes, it's probably better to support changing this through the CMake build options indeed. Sometimes people also disable KUBSAN for the x86 build, so they can use the build options then. :) 

> 3. Mostly unrelated to this patch, but explicit bit shuffling operations feel a bit more reliable to me than bitfields.

I definitely agree with you here, bitfields are basically undefined and differ per compiler, explicit bit shuffling is always defined. But for now the compiler seems to generate the correct instructions, so I'm fine with it. It would be better to change them at some point indeed!

File: Kernel/CMakeLists.txt (hunk 6 above)
Comment (kleinesfilmroellchen) [reply Kernel/CMakeLists.txt]: That's a lot of jank. If you need alignment, how about `max(sizeof(Mailbox::MessageHeader), max(sizeof(Mailbox::MessageTail), QueryFirmwareVersionMboxMessage))`, which is what the compiler should normally do? Or why is the align here in the first place, is that not done automatically?

File: Documentation/AdvancedBuildInstructions.md [hunk 7]
```diff
@@ -42,6 +42,7 @@ directory to `Build/i686` and then running `ninja <target>`:
 
 There are some optional features that can be enabled during compilation that are intended to help with specific types of development work or introduce experimental features. Currently, the following build options are available:
 - `ENABLE_ADDRESS_SANITIZER` and `ENABLE_KERNEL_ADDRESS_SANITIZER`: builds in runtime checks for memory corruption bugs (like buffer overflows and memory leaks) in Lagom test cases and the kernel, respectively.
+- `ENABLE_KERNEL_UBSAN`: builds in runtime checks for detecting undefined behavior in the kernel.
```
Comment (BertalanD) [inline Documentation/AdvancedBuildInstructions.md:45]: Nit, feel free to ignore: could we be consistent with the ASan/userspace UBSan flags and call it `ENABLE_KERNEL_UNDEFINED_SANITIZER`.

**Review by BertalanD — CHANGES_REQUESTED** [review]

File: Kernel/CMakeLists.txt (hunk 6 above)
Comment (nico) [reply Kernel/CMakeLists.txt]: I wonder if we had this problem with the LHS of https://github.com/SerenityOS/serenity/commit/3cc5752a69eebc1a92d447b3bf254faaa5eafb7b#diff-25f1ccd9541ff907401a9a8b70840177a86b0a670509641f7ea7cf28c4feef51L82 …

File: Documentation/AdvancedBuildInstructions.md (hunk 7 above)
Comment (FireFox317) [reply Documentation/AdvancedBuildInstructions.md]: Fixed :)

File: Kernel/CMakeLists.txt (hunk 6 above)
Comment (FireFox317) [reply Kernel/CMakeLists.txt]: @nico Guess what, we don't have the problem when using the old code. So that is definitely the issue!
